compiler/src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -488,12 +488,19 @@ class PostTyper extends MacroTransform with InfoTransformer { thisPhase =>
             //     case x: (_: Tree[?])
         case m @ MatchTypeTree(bounds, selector, cases) =>
           // Analog to the case above for match types
-          def transformIgnoringBoundsCheck(x: CaseDef): CaseDef =
-            withMode(Mode.Pattern)(super.transform(x)).asInstanceOf[CaseDef]
+          def transformCase(x: CaseDef): CaseDef =
+            val gadtCtx = x.pat.removeAttachment(typer.Typer.InferredGadtConstraints) match
+              case Some(gadt) => ctx.fresh.setGadtState(GadtState(gadt))
```

Comment:
Is this enough to ensure that, after `PostTyper`, the trees are well-kinded without resorting to GADT reasoning? I'm not familiar with what are the consequences of doing `setGadtState`.

Reply:
I'm not inserting any casts, if that's what you're asking.  And I'm hoping we can handle things like:

```scala
class Foo
class Bar
class Res[X <: Bar]
type MT[X <: Foo | Bar] = X match
  case Foo => Unit
  case Bar => Res[X] // X vs bounds check <: Bar 
```

in part also because I've found adding a type intersection (for types) doesn't fix errors like adding a type casting (for terms) fixes errors...

Reply:
So we can either:
1. allow these things to require GADT reasoning
2. deal with how to force types to honour bounds, and perhaps in a way that is accessible to users too
3. return to the AssumeInfo tree idea

😭